### PR TITLE
Fix inclusion header files and use forward declaration for API

### PR DESCRIPTION
**Pull request by cngzhnp**
Fixes #2043 but only API side.

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by cngzhnp** [comment]
@marcalff Could you please review these changes only related to API folder?

File: api/include/opentelemetry/plugin/hook.h [hunk 1]
```diff
@@ -41,6 +40,11 @@
 OPENTELEMETRY_BEGIN_NAMESPACE
 namespace plugin
 {
+
+struct LoaderInfo;
+class Factory;
```
Comment (lalitb) [inline api/include/opentelemetry/plugin/hook.h:45]: Is this required ?

File: api/include/opentelemetry/trace/noop.h [hunk 2]
```diff
@@ -102,15 +102,14 @@ class OPENTELEMETRY_EXPORT NoopTracer final : public Tracer,
 /**
  * No-op implementation of a TracerProvider.
  */
-class OPENTELEMETRY_EXPORT NoopTracerProvider final : public opentelemetry::trace::TracerProvider
+class OPENTELEMETRY_EXPORT NoopTracerProvider final : public trace_api::TracerProvider
```
Comment (lalitb) [inline api/include/opentelemetry/trace/noop.h:105]: should we instead use `trace::TracerProvider`, and remove `trace_api` namespace alias from the header file. It's not recommended to have namespace alias in header files (though this is existing code)

**Comment by lalitb** [comment]
Changes look good in general, and nicely done. I think this would be the breaking change for applications if they are taking dependency on these unnecessary header inclusions in our API. These applications need to explicitly include these removed header files / forward-declared classes.

**Review by lalitb — APPROVED** [review]
Approving the changes with nit comments. I believe @marcalff has already approved these api changes as indicated here - https://github.com/open-telemetry/opentelemetry-cpp/pull/2043#pullrequestreview-1395717324. Still have concern on this breaking the existing application code, so would be good to discuss when/how to merge/release in next community meeting.

File: api/include/opentelemetry/plugin/hook.h (hunk 1 above)
Comment (cngzhnp) [reply api/include/opentelemetry/plugin/hook.h:45]: No need to put this class as a forward declaration, probably even stands on my early trials. Will be removed.

File: api/include/opentelemetry/trace/noop.h (hunk 2 above)
Comment (cngzhnp) [reply api/include/opentelemetry/trace/noop.h:105]: Yes, you are right. I'd change it as you wish.

Comment (lalitb) [reply api/include/opentelemetry/trace/noop.h]: Looks like the `trace_api` alias defined here is been used in other places too. We can keep it here in that case.

Comment (cngzhnp) [reply api/include/opentelemetry/trace/noop.h]: Reverted back only `trace_api` alias. However, I will take it as a note, maybe it would be removed after same changes are applied to SDK too.

File: api/test/plugin/dynamic_load_test.cc [hunk 3]
```diff
@@ -2,6 +2,12 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "opentelemetry/plugin/dynamic_load.h"
+#ifdef _WIN32
```
Comment (ThomsonTan) [inline api/test/plugin/dynamic_load_test.cc:5]: Why is this needed as the `#ifdef _WIN32` is included in `opentelemetry/plugin/dynamic_load.h` which has already been included above?

Comment (cngzhnp) [reply api/test/plugin/dynamic_load_test.cc:5]: In `dynamic_load.h` file, there is no preprocessor definition anymore, I removed it in this pull request. Because there is only one function declaration which is not related to operating system itself. 

However, only this source file is related to operating system itself with dynamic_load and which has to be included like that. Otherwise, there will be a compiler error on test codes.

Comment (ThomsonTan) [reply api/test/plugin/dynamic_load_test.cc:5]: Why was the including of such `dynamic_load_<os>.h` moved out of `dynamic_load.h`? With change, the user code like the test code here need to include both files separately. It looks to me that including the os header in `dynamic_load.h` is more convenient.

Comment (cngzhnp) [reply api/test/plugin/dynamic_load_test.cc:5]: Ok, I will revert it back.

**Review by ThomsonTan — APPROVED** [review]

**Review by owent — APPROVED** [review]
